### PR TITLE
Implement kubernetes gitlab authentication script

### PR DESCRIPTION
**Pull request by targs08**
## Synopsis

Implement one page script, which receives [GitLab Personal Access Token](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html) and genereates kubeconfig using this token. 

## Solution

- Receive [GitLab Personal Access Token](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html) using `curl`
- Generate kubeconfig using `kubectl`

### CODE REVIEW

**Comment by targs08** [comment]
FCM

```
[kubernetes/gitlab-auth.sh] Create GitLab authentication Kubernetes CLI setup script (#6)
```

File: kubernetes/kube-gitlab-auth.sh [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+# getGitlabToken generate Gitlab personal access token with api scopes
+# Example:
+#   getGitlabToken "https://gitlab.com" "user" "pass"
+getGitlabToken() {
+  local GITLAB_URL=$1
+  local GITLAB_USER=$2
+  local GITLAB_PASS=$3
+  local GITLAB_TOKEN_NAME="kubernetes"
```
Comment (tyranron) [inline kubernetes/kube-gitlab-auth.sh]: @targs08 here we should specify the name of concrete cluster, not just `kubernetes`.

File: kubernetes/kube-gitlab-auth.sh [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+# runCmd prints the given command and runs it.
+runCmd() {
+  (set -x; $@)
+}
+
+# getGitlabToken generate Gitlab personal access token with api scopes
+# Example:
+#   getGitlabToken "https://gitlab.com" "user" "pass"
+getGitlabToken() {
```
Comment (tyranron) [inline kubernetes/kube-gitlab-auth.sh]: @targs08 `get` usually means receiving without modification. So it's better to name with `generate`.

File: kubernetes/kube-gitlab-auth.sh [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    kubectl config set-credentials gitlab.$gitlabUser \
+      --token $gitlabToken
+
+  for namespace in $(echo $ksClusterNamespaces | tr "," "\n")
+  do
+    runCmd \
+      kubectl config set-context $namespace \
+        --namespace=$namespace \
+        --cluster=$ksClusterName \
+        --user=gitlab.$gitlabUser
```
Comment (tyranron) [inline kubernetes/kube-gitlab-auth.sh]: @targs08 is `gitlab.` prefix mandatory? Can we just get rid of it?

**Review by tyranron — DISMISSED** [review]
Hard to read, the code should be more decoupled and simplified.

File: kubernetes/kube-gitlab-auth.sh (hunk 3 above)
Comment (targs08) [reply kubernetes/kube-gitlab-auth.sh]: @tyranron, not mandatory. Remove prefix

File: kubernetes/gitlab-auth.sh [hunk 4]
```diff
@@ -49,14 +49,13 @@ getGitlabToken() {
     | sed 's/.*<form class="new_user[^<]*\(<[^<]*\)\{2\}authenticity_token" value="\([^ ]*\)".*/\2/' \
     | sed -n 1p)
 
-  curl -b "$COOKIES_FILE" -c "$COOKIES_FILE" -s --output /dev/null \
+  curl -b "$COOKIES_FILE" -c "$COOKIES_FILE" -s --output=/dev/null \
     -i "$GITLAB_URL/users/sign_in" \
-    --data "user[login]=$GITLAB_USER&user[password]=$GITLAB_PASS" \
-    --data-urlencode "authenticity_token=$csrfToken"
+    --data="user[login]=$GITLAB_USER&user[password]=$GITLAB_PASS" \
```
Comment (targs08) [inline kubernetes/gitlab-auth.sh]: @tyranron, it won't work. [curl docs](https://curl.haxx.se/docs/manpage.html#-d)

**Comment by targs08** [comment]
@tyranron, what is your opinion, where should the code be simplified? 
The difficult place is `generateGitlabToken`. There is `html parse` and `large request data`. This function is base on [gist.github](https://gist.github.com/michaellihs/5ef5e8dbf48e63e2172a573f7b32c638) without `perl`. 

I don't see how this function can be improved. 😕 Sent to repeat `review`

Also, implement `verifyGitlabToken` function

**Review by tyranron — APPROVED** [review]
@targs08 I've corrected code style and fixed some minor errors (like `exit`ing from `verifyGitlabToken` rather than `return`ing).

Please, test the final result very carefully (check all possible conditions).

File: kubernetes/gitlab-auth.sh [hunk 5]
```diff
@@ -34,117 +34,116 @@ runCmd() {
...
+  local GITLAB_USER="$2"
+  local GITLAB_PASS="$3"
+  local GITLAB_TOKEN_NAME="$4"
+
+  local COOKIES_FILE="$(mktemp --suffix=cookies)"
 
+  # 1. Sign in into GitLab via username and password.
   local htmlContent=$(curl -c "$COOKIES_FILE" -i "$GITLAB_URL/users/sign_in" -s)
-  local csrfToken=$(echo $htmlContent \
+  local csrfToken=$(echo "$htmlContent" \
```
Comment (targs08) [inline kubernetes/gitlab-auth.sh]: @tyranron here and same places, quotes has broken functionality. I think cause of this is escapes symbols

**Comment by targs08** [comment]
@tyranron, some corrections and tested all conditions. Sent to repeat review

**Review by tyranron — APPROVED** [review]

**Pull request merged**